### ISSUE TITLE
natsConnection_Subscribe has error NATS_INVALID_SUBJECT

### ISSUE THREAD

**Issue by gtr32x** [nats-io/nats.c#254]
This is a first for me, and I cannot tell if I've compiled it incorrectly this time. I've hit two issues cannot link to Openssl and Protobuf-c while building so I linked the libraries against the brew installed paths to make it work. However I'm seeing this error `NATS_INVALID_SUBJECT` when doing `natsConnection_Subscribe`, which did not occur on another machine I had built. I literally put in the exact same string, cast it to `c_str()` from c++ codebase but having it fail.

Would definitely love any pointers here as to how to move on from here. Thank you!

I'm on Mac Mojave

**Comment by gtr32x** [comment]
So some additional information that I've discovered, previously it turned out that I was on `libnats.1.8`. And the new dylib that I've compiled is now `libnats.2.0.1`, has there been changes? Thanks.

**Comment by gtr32x** [comment]
So I actually suspect that it's this commit (6509a6baf61f6f681435357dc0baf9a40725662f ) that had altered the behavior. For what it's worth, my code looks like the following:

```
void subscribe(string channel, natsMsgHandler cb, void* handler)
    {
        natsStatus s;
        s = natsConnection_Subscribe(&_sub, _conn, channel.c_str(), cb, handler);
        if (s != NATS_OK)
            cleanAndThrow(s, "subscribe");
    }```

**Comment by gtr32x** [comment]
I haven't exactly spotted an issue in the code, but my input would be in the form of `T.ABCD`

**Comment by gtr32x** [comment]
Ok found the bug, `lastDot` should initialize to -1 instead of 0. Otherwise the dot at position 1 is perfectly valid.

**Comment by gtr32x** [comment]
Tried to do a patch but I don't think I have the proper access rights, here is what I would propose:

```diff --git a/src/conn.c b/src/conn.c
index bb35200..8bf46fe 100644
--- a/src/conn.c
+++ b/src/conn.c
@@ -2740,7 +2740,7 @@ _checkSubjOrQueue(const char *name, bool checkToken)
 {
     int     i       = 0;
     int     len     = 0;
-    int     lastDot = 0;
+    int     lastDot = -1;
     char    c;

     len = (int) strlen(name);
@@ -2752,7 +2752,7 @@ _checkSubjOrQueue(const char *name, bool checkToken)

         if (checkToken && (c == '.'))
         {
-            if ((i == 0) || (i == len-1) || (i == lastDot+1))
+            if ((i == len-1) || (i == lastDot+1))
                 return true;

             lastDot = i;
```

**Comment by kozlovic** [comment]
@gtr32x Thank you for the report and contribution! A PR is on its way.

I have added status in the past and forgot to bump the required header version. I had put in place a check `nats_CheckCompatibility()` that user can call in their application to make sure that the headers they were compiled with matches the dynamic library they are linked with. Unfortunately, I did not update the required version in the past releases. Not sure if you are using that function, but in the future, if you do and if I am updating properly, you would have noticed the difference. Sorry about that.
